### PR TITLE
Reverse KL with mask bug

### PR DESCRIPTION
**Pull request by oleksost**
# ✨ Description

Makign sure reverse KL loss is masked and averaged correctly.

## 🔍 Type of change

Select all that apply:

- [x] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [ ] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. Change A
2. Change B

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [ ] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/contributing/contributing).
- [ ] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [ ] 🎉 The functionality is complete, and I have tested the changes.
- [ ] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [ ] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration

- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing

- [ ] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

---

## 🗒️ Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

File: fast_llm/functional/entropy_loss.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def _fused_reverse_kl_base(
     # Compute loss terms: student_probs * log_ratio, then sum over vocab
     # This is equivalent to kl_div(..., log_target=True) but more memory efficient
     log_ratio = predicted_log_probability - target_log_probability
-    per_sample_loss = (predicted_probability * log_ratio).sum(dim=-1)
+    per_sample_loss = (predicted_probability * log_ratio).sum(dim=-1, keepdim=True)
```
Comment (jlamypoirier) [inline fast_llm/functional/entropy_loss.py:124]: I don't see how this changes anything, keepdim is the same as unsqueeze?

Comment (oleksost) [reply fast_llm/functional/entropy_loss.py:124]: The issue is with the reported loss, that grads are ok.

When we later multiply the loss by by the mask [here](https://github.com/ServiceNow/Fast-LLM/blob/3f97df706c04bd3dbbfeb29845f7fa265ac97181/fast_llm/functional/entropy_loss.py#L280), `per_sample_loss = per_sample_loss * loss_mask.unsqueeze(-1)`, per_sample_loss is [N] and mask is [N,1], so when multiplying those torch actually seem to left-pad the missing dims with 1, so it would be the same as if we did `per_sample_loss = per_sample_loss.unsqueeze(0) * loss_mask.unsqueeze(-1)` which results in N x N tensor.

Then taking the average of that is much larger than taking the average of `per_sample_loss * loss_mask`.

**Review by jlamypoirier — APPROVED** [review]
